### PR TITLE
Add flag for ignoring entries check pypi script

### LINKED ISSUE #342: add flag for ignoring entries with main in check pypi versions script

### LINKED ISSUE #414: Check pypi script not working with packages with only prerelease versions
Certain packages, for example "opentelemetry-semantic-conventions", only have prerelease versions. In this case, the script will try to run max() over an empty list and fail.

### PR DESCRIPTION
**Pull request by jonathan-eq**
Resolves #342
Resolves #414 

Depends on #386

### CODE REVIEW

File: komodo/check_up_to_date_pypi.py [hunk 1]
```diff
@@ -147,9 +148,14 @@ def run_check_up_to_date(
...
     propose_upgrade=False,
+    ignore=None,
 ):
     yaml = yaml_parser()
-    releases = load_from_file(yaml, release_file)
+    releases: dict = load_from_file(yaml, release_file)
+    if ignore:
+        for package_name, package_version in list(releases.items()):
+            if re.search(ignore, f"{package_name} {package_version}"):
+                del releases[package_name]
```
Comment (berland) [inline komodo/check_up_to_date_pypi.py:158]: An alternative way could be to do a dict comprehension, then you could avoid the [list(releases.items())](url) (it is usually an antipattern and some linters could erroneously suggest to remove the `list` function here)

File: komodo/check_up_to_date_pypi.py [hunk 2]
```diff
@@ -161,16 +167,20 @@ def run_check_up_to_date(
             insert_upgrade_proposals(upgrade_proposals_from_pypi, repository, releases)
             print(
                 "Writing upgrade proposals from pypi, "
-                "assuming nothing has changed with dependencies...",
+                "assuming nothing has changed with dependencies..."
             )
-            with open(propose_upgrade, mode="w", encoding="utf-8") as fout:
+            with open(propose_upgrade, "w") as fout:
```
Comment (berland) [inline komodo/check_up_to_date_pypi.py:172]: rebase mess again ? ;)

File: komodo/check_up_to_date_pypi.py [hunk 3]
```diff
@@ -147,9 +148,14 @@ def run_check_up_to_date(
...
     ),
     propose_upgrade=False,
+    ignore=None,
 ):
     yaml = yaml_parser()
-    releases = load_from_file(yaml, release_file)
+    releases: dict = load_from_file(yaml, release_file)
+    if ignore:
+        for package_name, package_version in list(releases.items()):
+            if re.search(ignore, f"{package_name} {package_version}"):
```
Comment (berland) [inline komodo/check_up_to_date_pypi.py:157]: What if we have some pypi package that includes "main" in its package name? We should probably be explicit whether the regexp refer to the version or name.

File: tests/test_up_to_date_pypi.py [hunk 4]
```diff
@@ -547,3 +547,130 @@ def side_effect(url: str, timeout):
...
+        maintainer: scout
+
+      main:
+        source: pypi
+        make: pip
+        maintainer: scout
+    """
+
+    with tmpdir.as_cwd():
+        with open("repository_file.yml", "w") as fout:
```
Comment (berland) [inline tests/test_up_to_date_pypi.py:613]: encoding="utf-8"...

Comment (berland) [reply tests/test_up_to_date_pypi.py:613]: Personally I prefer 

```
from pathlib import Path

Path("repository_file.yml").write_text(repository_file_contents, encoding="utf-8")
```

always over  `with open()`

See e.g. https://treyhunner.com/2018/12/why-you-should-be-using-pathlib/

File: komodo/check_up_to_date_pypi.py (hunk 1 above)
Comment (jonathan-eq) [reply komodo/check_up_to_date_pypi.py:158]: Either dict comprehension or dict(filter(filter_function, releases.items()))

File: tests/test_up_to_date_pypi.py (hunk 4 above)
Comment (jonathan-eq) [reply tests/test_up_to_date_pypi.py:613]: I might start using pathlib for working with paths, but it does not seem like such a big difference for writing or reading a file in the current directory.

File: komodo/check_up_to_date_pypi.py [hunk 5]
```diff
@@ -161,16 +167,20 @@ def run_check_up_to_date(
...
             write_to_file(repository, repository_file)
 
         major_upgrades, minor_upgrades, patch_upgrades, other_upgrades = [], [], [], []
         for name, versions in upgrade_proposals_from_pypi.items():
             pypi_latest = get_version.Version(versions["suggested"])
-            current_version = get_version.Version(versions["previous"])
+            try:
+                current_version = get_version.Version(versions["previous"])
+            except get_version.InvalidVersion:
+                print(f"Could not parse version {versions['previous']}")
```
[the remark this answers is not in the document]
Comment (jonathan-eq) [reply komodo/check_up_to_date_pypi.py:185]: Yes

[the remark this answers is not in the document]
Comment (berland) [reply komodo/check_up_to_date_pypi.py:185]: It's been too long to remember, but can this create noise in the output? `main` is an accepted version in bleeding.

[the remark this answers is not in the document]
Comment (jonathan-eq) [reply komodo/check_up_to_date_pypi.py:185]: Yes, but this only runs on packages which have pypi as a source in repository.yml. Main versions are sourced elsewhere, so it is filtered out at an earlier stage.

[the remark this answers is not in the document]
Comment (jonathan-eq) [reply komodo/check_up_to_date_pypi.py:185]: ^usually sourced with git

**Comment by jonathan-eq** [comment]
There is another issue with the code. It skips versions marked as pre-release, but certain packages only have pre-release versions. In this case, it will try to run max() over an empty list and fail.

**Review by berland — APPROVED** [review]
Nice improvement!

**Pull request merged**